README.md
# sky_device_info
A Flutter plugin to get device info.


## Usage
Quick simple usage example:

#### Read network info
```dart
final _plugin = SkyDeviceInfo();

DeviceInfo? deviceInfo = await _plugin.loadDeviceInfo();
if (deviceInfo != null) {
  NetworkInfo? networkInfo = _plugin.networkInfo;
  if (networkInfo != null && networkInfo.networkAdapters.isNotEmpty) {
    String? ipAddress = networkInfo.networkAdapters.first.ipAddress;
    String? connectionName =
        networkInfo.networkAdapters.first.connectionName;
    String? ssid = networkInfo.networkAdapters.first.ssid;
  } else {
    // Network disconnected
  }
}
```

<figure class="half">
    <img src="example_android.png" width="300px" alt="Screenshot_1" />
</figure>
<img src="example_windows.png" alt="Screenshot_2" />

CHANGELOG.md
## 0.0.7
Fix ios error.

## 0.0.6
Fix virtual network adapter.

## 0.0.5
Fix error.

## 0.0.4
Fix error.

## 0.0.3
Fix error.

## 0.0.2
Fix error.

## 0.0.1
Create project.

lib/beans.dart
import 'package:sky_device_info/utils.dart';

part 'beans.g.dart';

// @JsonSerializable()
class DeviceInfo {
  DeviceInfo({
    this.deviceName,
    this.osName,
    this.monitorName,
    this.gpu,
  });

  String? deviceName;
  String? osName;
  String? monitorName;
  List<GPUInfo>? gpu;

  factory DeviceInfo.fromJson(Map<String, dynamic> srcJson) =>
      _$DeviceInfoFromJson(srcJson);

  Map<String, dynamic> toJson() => filter(_$DeviceInfoToJson(this));
}

// @JsonSerializable()
class GPUInfo {
  GPUInfo({this.name, this.version});

  String? name;
  String? version;

  factory GPUInfo.fromJson(Map<String, dynamic> srcJson) =>
      _$GPUInfoFromJson(srcJson);

  Map<String, dynamic> toJson() => filter(_$GPUInfoToJson(this));
}

// @JsonSerializable()
class NetworkInfo {
  NetworkInfo({required this.networkAdapters});

  List<NetworkAdapter> networkAdapters;

  factory NetworkInfo.fromJson(Map<String, dynamic> srcJson) =>
      _$NetworkInfoFromJson(srcJson);

  Map<String, dynamic> toJson() => filter(_$NetworkInfoToJson(this));

  @override
  String toString() {
    return '{networkAdapters: $networkAdapters}';
  }
}

// @JsonSerializable()
class NetworkAdapter {
  final int ifTypeIEEE80211 = 71; // Radio spread spectrum
  String? name;
  String? id;
  int? index;
  int? bandwidth;
  String? connectionName;
  int? connectionStatus;
  String? ssid;
  bool? enabled;
  int? type; // 网卡类型
  String? macAddress; // 网卡MAC地址
  String? ipAddress; // IP 地址
  String? ipAddressIPv6; // IPv6地址
  String? ipMaskAddress; // 子网地址
  String? gatewayIpAddress; // 网关地址
  String? ipBroadcast; // 子网地址

  NetworkAdapter({
      this.name,
      this.id,
      this.index,
      this.bandwidth,
      this.connectionName,
      this.connectionStatus,
      this.ssid,
      this.enabled,
      this.type,
      this.macAddress,
      this.ipAddress,
      this.ipAddressIPv6,
      this.ipMaskAddress,
      this.gatewayIpAddress,
      this.ipBroadcast});

  factory NetworkAdapter.fromJson(Map<String, dynamic> srcJson) =>
      _$NetworkAdapterFromJson(srcJson);

  Map<String, dynamic> toJson() => filter(_$NetworkAdapterToJson(this));

  bool isWireless() => type == ifTypeIEEE80211;

  @override
  String toString() =>
      '{name: $name, id: $id, index: $index, bandwidth: $bandwidth,'
      ' connectionName: $connectionName, connectionStatus: $connectionStatus,'
      ' ssid: $ssid, enabled: $enabled, type: $type, macAddress: $macAddress,'
      ' ipAddress: $ipAddress, ipMaskAddress: $ipMaskAddress, gatewayIpAddress: $gatewayIpAddress}';
}


example/lib/main.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'dart:async';

import 'package:sky_device_info/beans.dart';
import 'package:sky_device_info/sky_device_info.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final _skyDeviceInfoPlugin = SkyDeviceInfo();
  String? _deviceInfo;
  String? _networkInfo;

  @override
  void initState() {
    super.initState();
    loadDeviceInfo();
    getIntranetIp();
  }

  @override
  void dispose() {
    super.dispose();
    _skyDeviceInfoPlugin.release();
  }

  String formatJson(Object? object) {
    return jsonEncode(object)
        .replaceAll('{', '{\n')
        .replaceAll('}', '\n}')
        .replaceAll(',', ',\n');
  }

  loadDeviceInfo() async {
    DeviceInfo? deviceInfo = await _skyDeviceInfoPlugin.loadDeviceInfo();
    setState(() {
      _deviceInfo = formatJson(deviceInfo ?? 'Unknown');
    });
    _skyDeviceInfoPlugin.addNetworkCallback((NetworkInfo? info) {
      NetworkInfo? networkInfo = _skyDeviceInfoPlugin.networkInfo;
      setState(() {
        _networkInfo = formatJson(networkInfo ?? 'Unknown');
      });
    });
    // _skyDeviceInfoPlugin.onNetworkChanged = (NetworkInfo? info) {
    //   NetworkInfo? networkInfo = _skyDeviceInfoPlugin.networkInfo;
    //   setState(() {
    //     _networkInfo = formatJson(networkInfo ?? 'Unknown');
    //   });
    // };
  }

  Future<String?> getIntranetIp() async {
    String? ip;
    for (var interface in await NetworkInterface.list()) {
      SkyDeviceInfo.log('interface ${interface.name} ${interface.index}');
      for (var addr in interface.addresses) {
        SkyDeviceInfo.log(
            'address ${addr.address} ${addr.host} ${addr.isLoopback} ${addr.isLinkLocal} ${addr.isMulticast}');
        if (addr.address.startsWith('192.') ||
            addr.address.startsWith('10.') ||
            addr.address.startsWith('172.')) ip = addr.address;
      }
    }
    return ip;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Container(
          padding: const EdgeInsets.all(20),
          child: Center(
            child:
                Text('DeviceInfo: $_deviceInfo\nNetworkInfo: $_networkInfo'),
          ),
        ),
      ),
    );
  }
}


windows/sky_device_info_plugin.h
#ifndef FLUTTER_PLUGIN_SKY_DEVICE_INFO_PLUGIN_H_
#define FLUTTER_PLUGIN_SKY_DEVICE_INFO_PLUGIN_H_

#include <flutter/method_channel.h>
#include <flutter/plugin_registrar_windows.h>

#include <memory>

namespace sky_device_info {

class SkyDeviceInfoPlugin : public flutter::Plugin {
 public:
  static void RegisterWithRegistrar(flutter::PluginRegistrarWindows *registrar);

  SkyDeviceInfoPlugin();

  virtual ~SkyDeviceInfoPlugin();

  // Disallow copy and assign.
  SkyDeviceInfoPlugin(const SkyDeviceInfoPlugin&) = delete;
  SkyDeviceInfoPlugin& operator=(const SkyDeviceInfoPlugin&) = delete;

 private:
  // Called when a method is called on this plugin's channel from Dart.
  void HandleMethodCall(
      const flutter::MethodCall<flutter::EncodableValue> &method_call,
      std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result);
};

}  // namespace sky_device_info

#endif  // FLUTTER_PLUGIN_SKY_DEVICE_INFO_PLUGIN_H_


ios/Classes/SkyDeviceInfoPlugin.m
#import "SkyDeviceInfoPlugin.h"

@implementation SkyDeviceInfoPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"sky_device_info"
            binaryMessenger:[registrar messenger]];
  SkyDeviceInfoPlugin* instance = [[SkyDeviceInfoPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else if ([@"loadDeviceInfo" isEqualToString:call.method]) {
    result([@"{\"deviceName\":\"" stringByAppendingString:[[UIDevice currentDevice] name]
      stringByAppendingString:@"\",\"osName\":\"iOS "
      stringByAppendingString:[[UIDevice currentDevice] systemVersion]
      stringByAppendingString:@"\"}"]);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end

/*
#import "SkyDeviceInfoPlugin.h"
#if __has_include(<sky_device_info/sky_device_info-Swift.h>)
#import <sky_device_info/sky_device_info-Swift.h>
#else
// Support project import fallback if the generated compatibility header
// is not copied when this plugin is created as a library.
// https://forums.swift.org/t/swift-static-libraries-dont-copy-generated-objective-c-header/19816
#import "sky_device_info-Swift.h"
#endif

@implementation SkyDeviceInfoPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  [SwiftSkyDeviceInfoPlugin registerWithRegistrar:registrar];
}
@end
*/

android/src/main/kotlin/com/ssnwt/sky_device_info/SkyDeviceInfoPlugin.kt
package com.ssnwt.sky_device_info

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.provider.Settings
import androidx.annotation.NonNull
import androidx.core.content.ContextCompat

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** SkyDeviceInfoPlugin */
class SkyDeviceInfoPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private var context: Context? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "sky_device_info")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.getApplicationContext()
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "loadDeviceInfo") {
      var deviceName = ""
      if (ContextCompat.checkSelfPermission(context!!, Manifest.permission.BLUETOOTH_CONNECT)
        === PackageManager.PERMISSION_GRANTED
      ) {
        deviceName = BluetoothAdapter.getDefaultAdapter().name
      } else {
        deviceName = Settings.Secure.getString(context?.getContentResolver(), "bluetooth_name")
      }
      var osName = "Android ${Build.VERSION.RELEASE}"//Build.DISPLAY
      result.success("{\"deviceName\":\"$deviceName\",\"osName\":\"$osName\"}")
    }
    if (call.method == "getPlatformVersion") {
      result.success("Android ${Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


lib/sky_device_info.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter/services.dart';
import 'package:sky_device_info/beans.dart';
import 'package:network_info_plus/network_info_plus.dart' as network_info_plus;

typedef OnNetworkChanged = Function(NetworkInfo?);

class SkyDeviceInfo {
  final _channel = const MethodChannel('sky_device_info');
  Function(DeviceInfo?)? _onDeviceInfo;
  DeviceInfo? _deviceInfo;
  NetworkInfo? _networkInfo;
  String? _networkInfoJson;
  final List<OnNetworkChanged> _onNetworkCallbackList = [];
  dynamic _networkSubscription;
  bool? _loadNetworkInfo;

  static log(Object? object) {
    bool inDebugMode = false;
    assert(inDebugMode = true);
    if (!inDebugMode) return;
    if (object == null) return;
    // ignore: avoid_print
    print('$object');
  }

  SkyDeviceInfo._privateConstructor();

  static final SkyDeviceInfo _instance = SkyDeviceInfo._privateConstructor();

  factory SkyDeviceInfo() {
    return _instance;
  }

  NetworkInfo? get networkInfo => _networkInfo;

  addNetworkCallback(OnNetworkChanged value) {
    _onNetworkCallbackList.add(value);
    if (_networkInfo != null) {
      value.call(_networkInfo);
    }
  }

  removeNetworkCallback(OnNetworkChanged value) {
    _onNetworkCallbackList.remove(value);
  }

  onNetworkChanged(NetworkInfo? networkInfo) {
    for (OnNetworkChanged item in _onNetworkCallbackList) {
      item.call(networkInfo);
    }
  }

  Future<NetworkInfo> readNetworkInfoByDart() async {
    List<NetworkAdapter> list = [];
    for (var interface in await NetworkInterface.list()) {
      // log('interface ${interface.name} ${interface.index}');
      for (var addr in interface.addresses) {
        // log('address ${addr.address} ${addr.host} ${addr.isLoopback}'
        //     ' ${addr.isLinkLocal} ${addr.isMulticast}');
        if (addr.address.startsWith('192.') ||
            addr.address.startsWith('10.') ||
            addr.address.startsWith('172.')) {
          NetworkAdapter adapter = NetworkAdapter(
              ipAddress: addr.address,
              connectionName: interface.name,
              index: interface.index);
          list.add(adapter);
        }
      }
    }
    return NetworkInfo(networkAdapters: list);
  }

  Future<NetworkInfo> _readNetworkInfo(connectivityResult) async {
    if (connectivityResult == ConnectivityResult.wifi) {
      final info = network_info_plus.NetworkInfo();
      // var wifiBSSID = await info.getWifiBSSID(); // 11:22:33:44:55:66
      String? ssid = await info.getWifiName();
      if (ssid != null &&
          ssid.length > 2 &&
          ssid.startsWith('"') &&
          ssid.endsWith('"')) {
        ssid = ssid.substring(1, ssid.length - 1);
      }
      NetworkAdapter adapter = NetworkAdapter(
          ssid: ssid,
          ipAddress: await info.getWifiIP(),
          ipAddressIPv6: await info.getWifiIPv6(),
          ipMaskAddress: await info.getWifiSubmask(),
          gatewayIpAddress: await info.getWifiGatewayIP(),
          ipBroadcast: await info.getWifiBroadcast(),
          index: 0);
      return NetworkInfo(networkAdapters: [adapter]);
    }
    return await readNetworkInfoByDart();
  }

  Future<DeviceInfo?> loadDeviceInfo({bool loadNetworkInfo = true}) async {
    if (_deviceInfo != null) return _deviceInfo!;
    _loadNetworkInfo = loadNetworkInfo;
    if (!Platform.isWindows) {
      String json = await _channel.invokeMethod('loadDeviceInfo');
      log('deviceInfo $json');
      _deviceInfo = DeviceInfo.fromJson(jsonDecode(json));

      if (loadNetworkInfo) {
        var connectivityResult = await Connectivity().checkConnectivity();
        log('connectivityResult $connectivityResult');
        _networkInfo = await _readNetworkInfo(connectivityResult);
        onNetworkChanged(_networkInfo);

        _networkSubscription = Connectivity()
            .onConnectivityChanged
            .listen((ConnectivityResult result) async {
          _networkInfo = await _readNetworkInfo(result);
          onNetworkChanged(_networkInfo);
        });
      }

      return _deviceInfo;
    }
    if (_onDeviceInfo != null) return null;
    Completer<DeviceInfo?> completer = Completer<DeviceInfo?>();
    _onDeviceInfo = (result) => completer.complete(result);

    _channel.setMethodCallHandler((MethodCall call) async {
      // print("playerPlugin method=${call.method} arguments=${call.arguments}");
      if (call.method == 'onDeviceInfo') {
        log('deviceInfo ${_onDeviceInfo != null} ${call.arguments}');
        DeviceInfo deviceInfo = DeviceInfo.fromJson(jsonDecode(call.arguments));
        _deviceInfo = deviceInfo;
        if (_loadNetworkInfo != true) {
          _onDeviceInfo?.call(deviceInfo);
          _onDeviceInfo = null;
        }
      } else if (call.method == 'onNetworkInfo') {
        if (_networkInfoJson != call.arguments) {
          _networkInfoJson = call.arguments;
          log('loadNetworkInfo ${call.arguments}');
          NetworkInfo networkInfo =
              NetworkInfo.fromJson(jsonDecode(call.arguments));
          _networkInfo = networkInfo;
          onNetworkChanged(_networkInfo);
        }
        if (_loadNetworkInfo == true) {
          _onDeviceInfo?.call(_deviceInfo);
          _onDeviceInfo = null;
        }
      }
    });

    final Map<String, dynamic> arguments = {
      'loadNetworkInfo': loadNetworkInfo,
    };
    _channel.invokeMethod('loadDeviceInfo', arguments);
    return completer.future;
  }

  release() {
    _channel.setMethodCallHandler(null);
    if (Platform.isWindows) {
      _channel.invokeMethod('release');
    }
    _networkSubscription?.cancel();
    _onNetworkCallbackList.clear();
    _deviceInfo = null;
    _networkInfo = null;
    _networkInfoJson = null;
  }
}
